Implementing http_client based on boost.beast and adding proxy support to tls logger.

uptycs-nishant opened the pull request:
Few points -
1. Replaced cpp-netlib as http client and implemented http_client based upon more raw beast interfaces.
2. keeping cpp-netlib as dependency due to uri parsing capability (beast does not support as of now).
3. Introducing proxy support to tls transport.
./osqueryd --help| grep proxy
  ...
   --proxy_hostname VALUE                    proxy hostname
  ...

theopolis commented:
ok to test

osqueryer commented:
:no_good: The commit 0e9a37fb703d65db1892c41df917dbafa683735b (Job results: [2348](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2348/)) failed the code audit or documentation test.

osqueryer commented:
:-1: The commit 0e9a37fb703d65db1892c41df917dbafa683735b (Job results: [5662](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5662/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 0e9a37fb703d65db1892c41df917dbafa683735b (Job results: [2548](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2548/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 0e9a37fb703d65db1892c41df917dbafa683735b (Job results: [2549](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2549/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 0e9a37fb703d65db1892c41df917dbafa683735b (Job results: [5663](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5663/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 0e9a37fb703d65db1892c41df917dbafa683735b (Job results: [605](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/605/)) failed one or more tests (**FreeBSD**).

osqueryer commented:
:-1: The commit 0e9a37fb703d65db1892c41df917dbafa683735b (Job results: [606](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/606/)) failed one or more tests (**FreeBSD**).

osqueryer commented:
:-1: The commit be5822438a569292a1464287e68b6a30ddb7593c (Job results: [607](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/607/)) failed one or more tests (**FreeBSD**).

osqueryer commented:
:-1: The commit be5822438a569292a1464287e68b6a30ddb7593c (Job results: [1776](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1776/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit be5822438a569292a1464287e68b6a30ddb7593c (Job results: [608](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/608/)) failed one or more tests (**FreeBSD**).

osqueryer commented:
:-1: The commit be5822438a569292a1464287e68b6a30ddb7593c (Job results: [2550](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2550/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit be5822438a569292a1464287e68b6a30ddb7593c (Job results: [2551](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2551/)) failed one or more tests (**Windows**).

theopolis commented:
Right, this will need FreeBSD build support and a Windows choco package for beast 'maybe'? I say maybe because I thought the library was include in boost 1.65 (which should be available for all platforms): http://www.boost.org/doc/libs/develop/libs/beast/doc/html/beast/introduction.html

uptycs-nishant commented:
Not clear from logs of Windows and FreeBSD where it is failing.

muffins commented:
For Windows we have a Powershell [build script for `boost-msvc`](https://github.com/facebook/osquery/blob/master/tools/provision/chocolatey/boost-msvc14.ps1), and we might just need to extend what libraries we're building, below is the error:
```
...
  Generating ../generated/tables_foreign/authorization_mechanisms.cpp
C:\jenkins\workspace\osqueryPullRequestWindows\TargetSystem\aws-windows-2016\include\osquery/http_client.h(16): fatal error C1083: Cannot open include file: 'boost/beast/core.hpp': No such file or directory [C:\jenkins\workspace\osqueryPullRequestWindows\TargetSystem\aws-windows-2016\build\windows10\osquery\distributed\osquery_distributed_plugins.vcxproj]
...
```
As for FreeBSD, I'm not entirely certain. The error is below:
```
...
In file included from /var/jenkins/workspace/osqueryPullRequestFreeBSD/TargetSystem/aws-freebsd-11/osquery/remote/transports/tls.h:14:
/var/jenkins/workspace/osqueryPullRequestFreeBSD/TargetSystem/aws-freebsd-11/include/osquery/http_client.h:16:10: fatal error: 'boost/beast/core.hpp' file not found
#include <boost/beast/core.hpp>
         ^
1 error generated.
gmake[4]: *** [osquery/logger/CMakeFiles/osquery_logger_plugins.dir/build.make:111: osquery/logger/CMakeFiles/osquery_logger_plugins.dir/plugins/tls_logger.cpp.o] Error 1
gmake[3]: *** [CMakeFiles/Makefile2:2139: osquery/logger/CMakeFiles/osquery_logger_plugins.dir/all] Error 2
[ 64%] Built target osquery_additional_amalgamation
...
```

osqueryer commented:
:-1: The commit dd4326265cc714e1e8805742cbd3851afa0f3389 (Job results: [5699](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5699/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit dd4326265cc714e1e8805742cbd3851afa0f3389 (Job results: [5700](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5700/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit dd4326265cc714e1e8805742cbd3851afa0f3389 (Job results: [2578](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2578/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit dd4326265cc714e1e8805742cbd3851afa0f3389 (Job results: [2579](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2579/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit d6d43ddb19a2eb6ebb86a5b69fdb41ccf16fda11 (Job results: [2580](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2580/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit d6d43ddb19a2eb6ebb86a5b69fdb41ccf16fda11 (Job results: [2581](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2581/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit d6d43ddb19a2eb6ebb86a5b69fdb41ccf16fda11 (Job results: [635](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/635/)) failed one or more tests (**FreeBSD**).

osqueryer commented:
:-1: The commit d6d43ddb19a2eb6ebb86a5b69fdb41ccf16fda11 (Job results: [640](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/640/)) failed one or more tests (**FreeBSD**).

osqueryer commented:
:-1: The commit 97bb9863e43838ba8424f81a90f079e68a410a3b (Job results: [2586](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2586/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 97bb9863e43838ba8424f81a90f079e68a410a3b (Job results: [2587](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2587/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 97bb9863e43838ba8424f81a90f079e68a410a3b (Job results: [641](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/641/)) failed one or more tests (**FreeBSD**).

osqueryer commented:
:-1: The commit 97bb9863e43838ba8424f81a90f079e68a410a3b (Job results: [642](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/642/)) failed one or more tests (**FreeBSD**).

theopolis commented:
@muffins it looks like the Beast boost project is not bundled in the main boost. For POSIX this requires a separate tgz to lay down the header-only library. We'll need the same for Windows most likely: https://github.com/boostorg/beast

uptycs-nishant commented:
beast is going to be part of Boost-1.66, till that time for this change we need to keep beast.ps1 for windows and beast.rb for linux/macos. I have tested this patch on windows using nupkg created using beast.ps1 (but need to upload that package). I not sure about freebsd, how to solve beast dependency problem on freebsd?

theopolis commented:
Is there no beast port for FreeBSD? If there isn't then we can gate it from FreeBSD until boost 1.66 using compiler and CMake defines. We do this with the SKIP_* environment variables, check those out for reference.

osqueryer commented:
:no_good: The commit d96d669aaa6fbdd864951e739f35f5dad4162131 (Job results: [2384](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2384/)) failed the code audit or documentation test.

uptycs-nishant commented:
I am done will all the code changes.
Added beast.ps1 for windows(beast package needs to be uploaded manually).
Modified freebsd.sh to fetch beast from uptycs-nishant fork of beast.

osqueryer commented:
:-1: The commit 12616744dfc9e492109ca3d3cf5c71f6ef49f3ae (Job results: [5712](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5712/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 12616744dfc9e492109ca3d3cf5c71f6ef49f3ae (Job results: [2593](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2593/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 12616744dfc9e492109ca3d3cf5c71f6ef49f3ae (Job results: [5713](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5713/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 12616744dfc9e492109ca3d3cf5c71f6ef49f3ae (Job results: [2594](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2594/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 8b3e2e06858be7e8f627d98d897340ac7d5e9ba9 (Job results: [2595](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2595/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 8b3e2e06858be7e8f627d98d897340ac7d5e9ba9 (Job results: [2596](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2596/)) failed one or more tests (**Windows**).

uptycs-nishant commented:
Please let me know if anything else is required.

osqueryer commented:
:-1: The commit b1fa1841963de2b4306e1eb26fe47c2c3dfcd8ce (Job results: [2603](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2603/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit b1fa1841963de2b4306e1eb26fe47c2c3dfcd8ce (Job results: [2604](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2604/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 4a5958575ffc5b1b65fcd8a110c28227d74ca9de (Job results: [2609](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2609/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 4a5958575ffc5b1b65fcd8a110c28227d74ca9de (Job results: [2610](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2610/)) failed one or more tests (**Windows**).

muffins commented on tools/provision/chocolatey/beast.ps1:
```diff
@@ -0,0 +1,109 @@
+#  Copyright (c) 2014-present, Facebook, Inc.
```
As this library is header only, we can trim it down quite a bit, would you mind making this script the following?
```
#  Copyright (c) 2014-present, Facebook, Inc.
#  All rights reserved.
#
#  This source code is licensed under the BSD-style license found in the
#  LICENSE file in the root directory of this source tree. An additional grant
#  of patent rights can be found in the PATENTS file in the same directory.

# Update-able metadata
#
# $version - The version of the software package to build
# $chocoVersion - The chocolatey package version, used for incremental bumps
#                 without changing the version of the software package

$version = '111.0'
$chocoVersion = $version
$packageName = 'beast'
$projectSource = 'https://github.com/boostorg/beast/tree/master'
$packageSourceUrl = 'https://github.com/boostorg/beast/tree/master'

$authors = 'boost'
$owners = 'boost'
$copyright = 'https://github.com/boostorg/beast/blob/master/LICENSE_1_0.txt'
$license = 'https://github.com/boostorg/beast/blob/master/LICENSE_1_0.txt'
$gitUrl = 'https://github.com/uptycs-nishant/beast.git'

# Invoke our utilities file
. "$(Split-Path -Parent $MyInvocation.MyCommand.Definition)\osquery_utils.ps1"

# Time our execution
$sw = [System.Diagnostics.StopWatch]::startnew()

# Keep the location of build script, to bring with in the chocolatey package
$buildScript = $MyInvocation.MyCommand.Definition

# Create the choco build dir if needed
$buildPath = Get-OsqueryBuildPath
if ($buildPath -eq '') {
  Write-Host '[-] Failed to find source root' -ForegroundColor red
  exit
}
$chocoBuildPath = "$buildPath\chocolatey\$packageName"
if (-not (Test-Path "$chocoBuildPath")) {
  New-Item -Force -ItemType Directory -Path "$chocoBuildPath"
}
Set-Location $chocoBuildPath

$sourceDir = Join-Path $(Get-Location) "$packageName"
$git = (Get-Command 'git').Source
$gitArgs = "clone $gitUrl"
Start-OsqueryProcess $git $gitArgs
Set-Location $sourceDir

# Checkout our working, patched, build of beast-v111
$gitArgs = 'checkout v111'
Start-OsqueryProcess $git $gitArgs

# Build the libraries, remove any old versions first.
$buildDir = Join-Path $(Get-Location) 'osquery-win-build'
if(Test-Path $buildDir){
  Remove-Item -Force -Recurse $buildDir
}
New-Item -Force -ItemType Directory -Path $buildDir
Set-Location $buildDir

# If the build path exists, purge it for a clean packaging
$chocoDir = Join-Path $(Get-Location) 'osquery-choco'
if (Test-Path $chocoDir) {
  Remove-Item -Force -Recurse $chocoDir
}

# Construct the Chocolatey Package
New-Item -ItemType Directory -Path $chocoDir
Set-Location $chocoDir
$includeDir = New-Item -ItemType Directory -Path 'local\include'
$srcDir = New-Item -ItemType Directory -Path 'local\src'

Write-NuSpec `
  $packageName `
  $chocoVersion `
  $authors `
  $owners `
  $projectSource `
  $packageSourceUrl `
  $copyright `
  $license

Copy-Item -Recurse "$buildDir\..\include\boost" $includeDir
Copy-Item $buildScript $srcDir
choco pack

Write-Host "[*] Build took $($sw.ElapsedMilliseconds) ms" `
  -ForegroundColor DarkGreen
if (Test-Path "$packageName.$chocoVersion.nupkg") {
  Write-Host `
    "[+] Finished building $packageName v$chocoVersion." `
    -ForegroundColor Green
}
else {
  Write-Host `
    "[-] Failed to build $packageName v$chocoVersion." `
    -ForegroundColor Red
}
```

muffins replied:
Once you've swapped that out, add the following
```
...
"aws-sdk-cpp.1.1.44",
"beast.111.0",
"boost-msvc14.1.65.0",
...
```
To the [Windows provision script](https://github.com/facebook/osquery/blob/master/tools/provision.ps1#L285). I've hosted the package in our S3 bucket, so if you make the changes and re-push, it should trigger the windows build and pass tests.

muffins commented:
It looks like you might need a rebase? We had another PR merge in this morning abstracting some of the `aws_util` functionality, after that I'd say this LGTM. I'm curious if @theopolis wants to give this another look through?

theopolis commented:
This may prevent us from releasing new versions for FreeBSD, since there is no port for `beast`.

osqueryer commented:
:-1: The commit 362459d89f2e9638025dc264d773438ae493dbc0 (Job results: [1851](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1851/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit 362459d89f2e9638025dc264d773438ae493dbc0 (Job results: [1852](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1852/)) failed one or more tests (**macOS/OS X**).

uptycs-nishant commented:
It is also being built on freebsd. Please check freebsd.sh for fetch_beast.

theopolis commented:
Right, but we include an osquery port in the FreeBSD ports tree. We need to keep that updated as we fix bugs and add features. The solution here while waiting for boost 1.66 is to add a beast port. Maybe @lxcode can lend us another favor :P

theopolis commented:
Ultimately this will allow us to remove Cpp-Netlib and the non-boost ASIO formula, so I'm really excited about porting to beast!

uptycs-nishant commented:
boost.beast is header only library and that is required only while compilation of osquery, I took advantage of this fact and made changes in freebsd.sh in the form of fetch_beast. 
In December when we have boost-1.66 we can get rid of beast.rb, beast.ps1 and fetch_beast (from freebsd.sh).

osqueryer commented:
:-1: The commit 6e0d0f74e3c2aa5b55fbe8f08d77b5d1d9433f43 (Job results: [1887](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1887/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit 6e0d0f74e3c2aa5b55fbe8f08d77b5d1d9433f43 (Job results: [5779](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5779/)) failed one or more tests (**Linux**).

osqueryer commented:
:no_good: The commit 32db0038ed6a1f50338a9dac270b685dafde1cff (Job results: [2452](https://jenkins.osquery.io/job/osqueryPullRequestAudit/2452/)) failed the code audit or documentation test.

osqueryer commented:
:-1: The commit 32db0038ed6a1f50338a9dac270b685dafde1cff (Job results: [1902](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1902/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit c3e584bcb1ede0aca24fdbd0aa8b1918a57aab0b (Job results: [5794](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5794/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit c3e584bcb1ede0aca24fdbd0aa8b1918a57aab0b (Job results: [1905](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1905/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit c3e584bcb1ede0aca24fdbd0aa8b1918a57aab0b (Job results: [5795](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5795/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit c3e584bcb1ede0aca24fdbd0aa8b1918a57aab0b (Job results: [1906](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1906/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit c3e584bcb1ede0aca24fdbd0aa8b1918a57aab0b (Job results: [721](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/721/)) failed one or more tests (**FreeBSD**).

osqueryer commented:
:-1: The commit c3e584bcb1ede0aca24fdbd0aa8b1918a57aab0b (Job results: [722](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/722/)) failed one or more tests (**FreeBSD**).

osqueryer commented:
:-1: The commit c3e584bcb1ede0aca24fdbd0aa8b1918a57aab0b (Job results: [2676](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2676/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit c3e584bcb1ede0aca24fdbd0aa8b1918a57aab0b (Job results: [2680](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2680/)) failed one or more tests (**Windows**).

uptycs-nishant commented:
So, captain ( @theopolis  ) we will not be sailing unless we have the port(beast) clearance ? :)

theopolis commented:
Nope, we had some sage advice from @lxcode, we'll place the header into `third-party` in a follow up commit. :)

theopolis commented on docs/wiki/deployment/remote.md:
```diff
@@ -9,13 +9,16 @@ The most important differentiator to the **filesystem** suite of plugins is an a
-3. Submit an `--enroll_secret_path`, an `--enroll_secret_env`, or use TLS-client authentication, to the enroll endpoint.
-4. Receive a **node_key** and store within the node's persistent storage (RocksDB).
-5. Make config/logger requests while providing **node_key** as identification/authentication.
+2. Configure a proxy `--proxy_hostname`(Optional Step).
```
nit, space between `--proxy_hostname` and (

uptycs-nishant replied:
ok.

theopolis commented on docs/wiki/deployment/remote.md:
```diff
@@ -9,13 +9,16 @@ The most important differentiator to the **filesystem** suite of plugins is an a
-4. Receive a **node_key** and store within the node's persistent storage (RocksDB).
-5. Make config/logger requests while providing **node_key** as identification/authentication.
+2. Configure a proxy `--proxy_hostname`(Optional Step).
+3. Place your server's root certificate authority's PEM-encoded certificate into a file, for example `/path/to/server-root.pem` and configure the note client to pin this root: `--tls_server_certs=`.
```
do you mind updating "the client to pin to these roots" since you'll be the new blame ;)

uptycs-nishant replied:
ok.

theopolis commented on include/osquery/http_client.h:
```diff
@@ -0,0 +1,368 @@
+namespace boost_asio = boost::asio;
+namespace beast_http = boost::beast::http;
+
+typedef boost_asio::ssl::stream<boost_asio::ip::tcp::socket&> ssl_stream;
```
Can we put these `typedef`s into the `osquery::http` namespace?

uptycs-nishant replied:
ok.

theopolis commented on include/osquery/http_client.h:
```diff
@@ -0,0 +1,368 @@
+ * These methods take request as refrence and  return respose by value.
+ */
+class Client {
+ public: // struct(s)
```
What is `// struct(s)` for?

uptycs-nishant replied:
ok.

theopolis commented on include/osquery/http_client.h:
```diff
@@ -0,0 +1,368 @@
+    friend class Client;
+  };
+
+ public: // methods
```
We don't need to name each, struct list/methods/etc.

uptycs-nishant replied:
ok.

theopolis commented on include/osquery/http_client.h:
```diff
@@ -0,0 +1,368 @@
+  }
+
+ private:
+  T* resp_;
```
Could this be a `std::unique_ptr`?

uptycs-nishant replied:
No, since I am not handing over the ownership.

vinniefalco replied:
I'm not sure why this class even exists. As far as I can tell it is to make the client calling code see `std::string` instead of `string_view`. But this comes at significant run-time cost.

theopolis commented on osquery/remote/http/http_client.cpp:
```diff
@@ -0,0 +1,317 @@
+namespace osquery {
+namespace http {
+
+/// This class is used to convert boost::system_exception
```
Can you use the long-form docblock?

uptycs-nishant replied:
ok.

theopolis commented on osquery/remote/http/http_client.cpp:
```diff
@@ -0,0 +1,317 @@
+  }
+
+ private:
+  boost::system::error_category const* pc_;
```
Can this be a unique pointer?

uptycs-nishant replied:
No, since I am not handing over the ownership.

theopolis commented on osquery/remote/http/http_client.cpp:
```diff
@@ -0,0 +1,317 @@
+    default:
+      return Response(resp.release());
+    }
+  } while (true);
```
Is this request respecting a timeout? To prevent an infinite loop here?

uptycs-nishant replied:
if client_options_.timeout_ is set then request times out see below -  
boost_asio::deadline_timer timer{ios_};
  if (client_options_.timeout_) {
    timer.expires_from_now(
        boost::posix_time::seconds(client_options_.timeout_));
    timer.async_wait(
        [=](boost_system::error_code const& ec) { timeoutHandler(ec); });
  }

uptycs-nishant replied:
Good point. Yes, there is a chance of infinite loop here in case of if some malicious website continuously sends back the redirect in response. Earlier I had put individual aync_read under timeout. Now I have put the entire http request under timeout.

vinniefalco replied:
I think you need to put a limit on the total number of redirections:
https://stackoverflow.com/questions/10895406/what-is-the-maximum-number-of-http-redirections-allowed-by-all-major-browsers

uptycs-nishant replied:
After studying this issue, I went ahead with timeout solution.

uptycs-nishant commented:
Beast does not provide uri parsing. We need to depend upon cpp-netlib for that or any other library (e.g. https://github.com/reBass/uri) or write our own uri parser.

osqueryer commented:
:-1: The commit 33167ecab9b6045e090e730744840200096672db (Job results: [5824](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5824/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 33167ecab9b6045e090e730744840200096672db (Job results: [5825](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5825/)) failed one or more tests (**Linux**).

theopolis commented:
Thanks for the changes @uptycs-nishant, we're ready to land this but we're going to wait for `2.8.0` to be marked as stable. If our internal deployment of that tag goes well we'll move this into `master` then immediately handle the "vendoring" of `beast` into FreeBSD such that we are ready for a `2.9.0` or equiv.

uptycs-nishant commented:
Aye aye, Captain ( @theopolis ). :)

vinniefalco commented on include/osquery/http_client.h:
```diff
@@ -0,0 +1,438 @@
+#include <boost/asio/ssl.hpp>
+#include <boost/beast/core.hpp>
+#include <boost/beast/http.hpp>
+#include <boost/optional/optional.hpp>
```
I would try to include only what you need instead of including `<boost/asio.hpp>` and `<boost/beast/core.hpp>`, otherwise you might find compilation times growing uncomfortably.

uptycs-nishant replied:
These two headers are required to compile the code successfully.

vinniefalco commented on include/osquery/http_client.h:
```diff
@@ -0,0 +1,438 @@
+namespace http {
+
+typedef boost_asio::ssl::stream<boost_asio::ip::tcp::socket&> ssl_stream;
+typedef beast_http::request<beast_http::string_body> beast_http_request;
```
Consider `empty_body` instead of `string_body`, since the requests never have a payload (or do they?)

vinniefalco commented on include/osquery/http_client.h:
```diff
@@ -0,0 +1,438 @@
+/**
+ * @brief a simple http client class based upon boost.beast.
+ *
+ * Implements put, post, get, head and delete_ methods.
```
minor typo, replace `delete_` with `delete`

uptycs-nishant replied:
That was intentional, I was listing the actual method. 
But it makes more sense to change it to delete.

vinniefalco commented on include/osquery/http_client.h:
```diff
@@ -0,0 +1,438 @@
+ * @brief a simple http client class based upon boost.beast.
+ *
+ * Implements put, post, get, head and delete_ methods.
+ * These methods take request as refrence and  return respose by value.
```
Typo: `refrence`, `respose`

uptycs-nishant replied:
corrected.

vinniefalco commented on include/osquery/http_client.h:
```diff
@@ -0,0 +1,438 @@
+   * function 'operator<<' of HTTP_Request class.
+   */
+  struct Header {
+    Header(const std::string& name, const std::string& value)
```
Consider using these constructor signatures instead:
```
Header(beast_http::field name, boost::beast::string_param value);
Header(boost::beast::string_view name, boost::beast::string_param value);
```

vinniefalco replied:
Actually I would just remove the `Header` class entirely, it doesn't seem to add any benefit that I can tell, and in most cases introduces unnecessary dynamic memory allocations.

vinniefalco commented on include/osquery/http_client.h:
```diff
@@ -0,0 +1,438 @@
+  };
+
+ public:
+  HTTP_Request(const std::string& url = std::string()) : uri_(url) {}
```
You might as well set the HTTP version in the constructor to 11 as well

uptycs-nishant replied:
point is noted, but if you explicitly specify then you don't have to comment about the default behavior.

vinniefalco commented on include/osquery/http_client.h:
```diff
@@ -0,0 +1,438 @@
+ * @brief HTTP response class
+ *
+ * This class is inherited from implementation(boost.beast) http response class.
+ * This class gives convinent access to some functionality of implemention
```
Spelling: `convenient`

uptycs-nishant replied:
corrected

vinniefalco commented on include/osquery/http_client.h:
```diff
@@ -0,0 +1,438 @@
+template <typename T>
+class HTTP_Response : public T {
+ public:
+  HTTP_Response() {}
```
`HTTP_Response() = default;`

uptycs-nishant replied:
Still stuck in cpp std2003 era. 
corrected.

vinniefalco commented on include/osquery/http_client.h:
```diff
@@ -0,0 +1,438 @@
+  }
+
+  /// body of an http response
+  const std::string& body() {
```
In the most recent version of Beast, `body()` is a member function of `message` so you might not need this.

uptycs-nishant replied:
Will look into it. Better for me lesser code.

uptycs-nishant replied:
We are at v111. Did not find body() method.

vinniefalco replied:
`body()` is in version 116:
https://github.com/boostorg/beast/commit/54fe7cacf75fd33f8e98d79f39f0214ac972beae

vinniefalco commented on include/osquery/http_client.h:
```diff
@@ -0,0 +1,438 @@
+
+  /// status of an http response
+  unsigned status() {
+    return this->T::result_int();
```
What is the purpose of having `status()`? Is it to make existing code compile?

uptycs-nishant replied:
you got it right.

vinniefalco commented on include/osquery/http_client.h:
```diff
@@ -0,0 +1,438 @@
+  /// All headers of an http response.
+  /// Headers can be accesed via HTTP_Response<T>::Iterator class
+  Headers headers() {
+    return Headers(this);
```
I'm not sure why you need this, `beast::http::fields` is already an iterable container. What is this for?

vinniefalco commented on include/osquery/http_client.h:
```diff
@@ -0,0 +1,438 @@
+
+  auto operator-> () {
+    return std::make_shared<std::pair<std::string, std::string>>(
+        std::string(iter_->name_string()), std::string(iter_->value()));
```
This is going to be very painful. Every time you use `operator->` you are performing 3 dynamic allocations! Beast took great care to avoid dynamic allocations when accessing fields. Why not just return `std::pair<beast::string_view, beast::string_view>` by value and avoid the dynamic allocation?

vinniefalco commented on include/osquery/http_client.h:
```diff
@@ -0,0 +1,438 @@
+
+  std::string operator[](const std::string& name) {
+    return std::string(resp_->T::operator[](name));
+  }
```
Why is this member function necesssary? It forces the caller to allocate a string for the `name` argument and it forces another allocation to construct `std::string` from `string_view`. This member function should be deleted.

vinniefalco commented on osquery/remote/http/http_client.cpp:
```diff
@@ -0,0 +1,342 @@
+  if ((ec.category() == boost_asio::error::ssl_category) &&
+      (ec.value() == SHORT_READ_ERROR)) {
+    // ignoring short read error
+    ec_ = boost_system::errc::make_error_code(boost_system::errc::success);
```
This call is unnecessarily expensive since it will cross a memory fence to retrieve the global error category. Use this to clear the error instead:
```
ec_.reset(0, ec_.category());
```

uptycs-nishant replied:
I believe you mean to say. ec_.clear() (reset does not exists). 

The way I see is, I will be saving a copy operation at the loss of readability.

error_code()  : m_val(0), m_cat(&system_category()) {}
VS
inline error_code make_error_code( errc_t e )                                                                                                        
   { return error_code( e, generic_category() );

vinniefalco replied:
My mistake, the correct code is
```
ec.assign(0, ec.category());
```
This is definitely the best way to set the error to "success" - I learned this technique from the Asio author himself!

uptycs-nishant replied:
now I am using ec_.clear() and have put a comment  above it.

vinniefalco commented on osquery/remote/http/http_client.cpp:
```diff
@@ -0,0 +1,342 @@
+  if ((ec.category() == boost_asio::error::ssl_category) &&
+      (ec.value() == SHORT_READ_ERROR)) {
+    // ignoring short read error
+    ec_ = boost_system::errc::make_error_code(boost_system::errc::success);
```
I have strong reservations about ignoring the "short read" error. In fact I am certain that this leads to a security vulnerability known as a "truncation attack", described here:
https://en.wikipedia.org/wiki/Transport_Layer_Security#Truncation_attack

See:
https://github.com/cpp-netlib/cpp-netlib/issues/670
https://stackoverflow.com/questions/25587403/boost-asio-ssl-async-shutdown-always-finishes-with-an-error

uptycs-nishant replied:
Your point is valid, but you can see from the code. We are closing the socket once we receive the response.

vinniefalco commented on osquery/remote/http/http_client.cpp:
```diff
@@ -0,0 +1,342 @@
+    // ignoring short read error
+    ec_ = boost_system::errc::make_error_code(boost_system::errc::success);
+  } else if ((ec.value() != boost_system::errc::operation_canceled) ||
+             (ec.category() != boost_asio::error::system_category)) {
```
You should never have to call `value()` or `category()` under normal usage, the following should work:
```
} else if (ec != boost_system::errc::operation_canceled) {
```

vinniefalco commented on osquery/remote/http/http_client.cpp:
```diff
@@ -0,0 +1,342 @@
+    boost::beast::flat_buffer b;
+    beast_http_response_parser rp;
+    rp.skip(true);
+    beast_http::read_header(sock_, b, rp);
```
What if the response contains "Content-Length: 1000"? This code will happily accept it, but in fact it is an error according to rfc7230. Consider using `beast_http::response<beast_http::empty_body>` instead, and call `read` instead of `read_header`.  The parser will generate an error if the message indicates there is any payload.

uptycs-nishant replied:
This code part it contacting the proxy server. Even if we receive the 1000 bytes we still want to connect to proxy.

vinniefalco replied:
It depends on how strict you want your implementation to adhere to rfc7230. If the proxy server is not under your control you might want to be more strict.

vinniefalco commented on osquery/remote/http/http_client.cpp:
```diff
@@ -0,0 +1,342 @@
+    beast_http_response_parser rp;
+    rp.skip(true);
+    beast_http::read_header(sock_, b, rp);
+    if (rp.get().result() != beast_http::status::ok) {
```
I think you have to consider any 2xx response as "successful" according to rfc7230, consider this instead:
```
if(beast_http::to_status_class(rp.get().result()) != beast_http::status_class::successful)
```

uptycs-nishant replied:
done.

vinniefalco commented on osquery/remote/http/http_client.cpp:
```diff
@@ -0,0 +1,342 @@
+                         beast_http_response_parser& resp) {
+  req.target((req.remotePath()) ? *req.remotePath() : "/");
+  req.version = 11;
+  req << Request::Header("Host", *client_options_.remote_hostname_);
```
I'm struggling to see the value of this syntax, instead of:
```
req.set(beast_http::field::host, *client_options_.remote_hostname_);
```
The less code, the better, `Request::Header` seems unnecessary.

uptycs-nishant replied:
Done. Within the boundaries of http_client , I can use the direct calls of beast.

vinniefalco commented on osquery/remote/http/http_client.cpp:
```diff
@@ -0,0 +1,342 @@
+    }
+
+    if ((ec.value() != boost_system::errc::operation_canceled) ||
+        (ec.category() != boost_asio::error::system_category)) {
```
This is unnecessary, you can just write:
```
if(ec != boost::asio::operation_aborted)
```

vinniefalco commented on osquery/remote/http/http_client.cpp:
```diff
@@ -0,0 +1,342 @@
+        postResponseHandler(ec);
+      });
+
+  ios_.run();
```
FYI: If an asynchronous operation throws an exception, the `io_service` will not be reset.

uptycs-nishant replied:
Two reasons - 
1. In our usage we use the client object once. 
2. I find there is no reason for this to throw the exception in this context.

I will keep this in mind once we enhance the http_client class in the future.

vinniefalco commented on osquery/remote/http/http_client.cpp:
```diff
@@ -0,0 +1,342 @@
+  req.method(beast_http::verb::put);
+  req.body = body;
+  if (content_type.size()) {
+    req << Request::Header("Content-Type", content_type);
```
Consider using this instead:
```
if(! content_type.empty())
    req.set(beast_http::field::content_type, content_type);
```

uptycs-nishant replied:
done.

vinniefalco commented on osquery/remote/http/http_client.cpp:
```diff
@@ -0,0 +1,342 @@
+                     std::string const& body,
+                     std::string const& content_type) {
+  req.method(beast_http::verb::put);
+  req.body = body;
```
Consider this code instead:
```
Response
Client::put(
    Request& req,
    std::string body,
    boost::beast::string_view content_type)
{
    req.method(beast_http::verb::put);
    req.body = std::move(body);
...
```

uptycs-nishant replied:
Appears to be, we are still copying body.

Anyway I have introduced an overload function as well -
  /// HTTP post request method with rvalue reference to body param
Response Client::post(Request& req,
                      std::string&& body,
                      std::string const& content_type) {
  req.method(beast_http::verb::post);
  req.body = std::move(body);
  if (content_type.size()) {
    req.set(beast_http::field::content_type, content_type);
  }
  return sendHTTPRequest(req);
}

vinniefalco commented on osquery/remote/transports/tls.cpp:
```diff
@@ -77,18 +79,22 @@ TLSTransport::TLSTransport() : verify_peer_(true) {
-  r << boost::network::header("Host", FLAGS_tls_hostname);
-  r << boost::network::header("User-Agent", kTLSUserAgentBase + kVersion);
+void TLSTransport::decorateRequest(http::Request& r) {
+  r << http::Request::Header("Connection", "close");
```
Consider using this instead:
```
r.keep_alive(false);
```

vinniefalco commented on osquery/remote/transports/tls.cpp:
```diff
@@ -77,18 +79,22 @@ TLSTransport::TLSTransport() : verify_peer_(true) {
+  r << http::Request::Header("Content-Type", serializer_->getContentType());
+  r << http::Request::Header("Accept", serializer_->getContentType());
+  r << http::Request::Header("Host", FLAGS_tls_hostname);
+  r << http::Request::Header("User-Agent", kTLSUserAgentBase + kVersion);
```
Consider using this code instead:
```
r.set(beast_http::field::content_type, serializer_->getContentType());
r.set(beast_http::field::accept, serializer_->getContentType());
r.set(beast_http::field::host, FLAGS_tls_hostname);
r.set(beast_http::field::user_agent, kTLSUserAgentBase + kVersion);
```

vinniefalco commented on osquery/tables/networking/posix/curl.cpp:
```diff
@@ -28,22 +26,22 @@ const std::string kOsqueryUserAgent{"osquery"};
 
     // Change the user-agent for the request to be osquery
-    request_ << header("User-Agent", r["user_agent"]);
+    request_ << osquery::http::Request::Header("User-Agent", r["user_agent"]);
```
Consider this instead:
```
request_.set(beast_http::field::user_agent, r["user_agent"]);
```

vinniefalco commented on osquery/tables/networking/posix/curl.cpp:
```diff
@@ -28,22 +26,22 @@ const std::string kOsqueryUserAgent{"osquery"};
     time_point<system_clock> end = std::chrono::system_clock::now();
 
-    r["response_code"] = INTEGER(static_cast<int>(status(response_)));
+    r["response_code"] = INTEGER(static_cast<int>(response_.status()));
```
Consider this instead:
```
r["response_code"] = response_.result_int();
```

And remove the `Response::status` member function

vinniefalco commented on osquery/tables/networking/posix/curl.cpp:
```diff
@@ -28,22 +26,22 @@ const std::string kOsqueryUserAgent{"osquery"};
     r["round_trip_time"] =
         BIGINT(duration_cast<microseconds>(end - start).count());
-    r["result"] = static_cast<std::string>(body(response_));
+    r["result"] = response_.body();
```
Consider this instead:
```
r["result"] = std::move(response_.body); // older Beast version
```
Alternatively, if you are using the latest Beast version:
```
r["result"] = std::move(response_.body());
```
And remove the member function `Response::body`

vinniefalco commented on osquery/utils/aws_util.cpp:
```diff
@@ -135,11 +117,11 @@ std::shared_ptr<Aws::Http::HttpResponse> NetlibHttpClient::MakeRequest(
 
   for (const auto& requestHeader : request.GetHeaders()) {
-    req << bn::header(requestHeader.first, requestHeader.second);
+    req << http::Request::Header(requestHeader.first, requestHeader.second);
```
Consider using:
```
req.set(requestHeader.first, requestHeader.second);
```

vinniefalco commented on osquery/utils/aws_util.cpp:
```diff
@@ -368,14 +340,13 @@ void getInstanceIDAndRegion(std::string& instance_id, std::string& region) {
     try {
-      http::client::response res = client.get(req);
+      http::Response res = client.get(req);
       if (res.status() == 200) {
```
As before, I think you need to check accept all 2xx values, like this:
```
if(beast_http::to_status_class(res.result()) != beast_http::status_class::successful)
```
And remove the member function `Response::status`

vinniefalco commented on osquery/utils/aws_util.cpp:
```diff
@@ -418,13 +389,13 @@ bool isEc2Instance() {
     try {
-      http::client::response res = client.get(req);
+      http::Response res = client.get(req);
       if (res.status() == 200) {
```
```
if(beast_http::to_status_class(res.result()) != beast_http::status_class::successful)
```

vinniefalco commented:
I'm always glad to see new Beast users, and it looks like this pull request implements a solid HTTP client which handles proxies and redirects. I've left code review comments on everything I could find. While the code should work (except for the one vulnerability with respect to handling SSL short reads) I think that a bit of extra polish would help trim it down, make it more readable and idiomatic, and also eliminate a lot of unnecessary dynamic memory allocation as I have indicated in the comments.

Side note: The flood of automated messages from @osqueryer and @facebook-github-bot make it hard to read the pull request comments. I wonder if they are adding enough value to justify the spam.

Good luck!

uptycs-nishant commented:
@vinniefalco , thanks for review comments  

I wrote http_client keeping following points in my mind -
1. Hide the implementation(boost.beast) details in couple of files http_client.{cpp,h}
2. Remove the cpp-netlib dependency and compile the code(osquery) and remove the compilation errors via writing new apis without touching(as much as possible) the user code. 
3. Shield the user code from changing apis of beast.

uptycs-nishant commented on include/osquery/http_client.h:
```diff
@@ -0,0 +1,438 @@
+class HTTP_Response : public T {
+ public:
+  HTTP_Response() {}
+  HTTP_Response(const T& resp) : T(resp) {}
```
One comment on my own code  :)
need to change constructor to - 
HTTP_Response(T&& resp) : T(std::move(resp)) {} 
in-order to achieve zero-copy here  "Response(resp.release())"

osqueryer commented:
:-1: The commit 49be2a48167dd67e2bb8f5a24ea8382fa66407f5 (Job results: [1954](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1954/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit 49be2a48167dd67e2bb8f5a24ea8382fa66407f5 (Job results: [767](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/767/)) failed one or more tests (**FreeBSD**).

theopolis commented:
Thanks @vinniefalco! You rock!

The for the side notes, I absolutely hear you loud and clear. The Jenkins/osqueryer bot provides very useful audit trails of failing jobs. Without its comments it's difficult to discern the association of historical job failures in Jenkins to GitHub pull requests. We find that we often refer to parts of historic job failures during pull request conversations and issues. The Facebook GitHub bot also contributes to auditing. If an author force-updates their branch we find it very difficult to reference or view the overwritten commit data.

I'm always open to recommendations and examples on how to provide this useful auditing without incurring GitHub user experience degradation!

vinniefalco commented:
>If an author force-updates their branch we find it very difficult to reference or view the overwritten commit data.

Yep, that is my experience as well. There are no perfect solutions, only finding a balance between the least-bad choices.

vinniefalco commented:
If anything, this pull request has been a wake-up call for me. I need to get crackin' on my uri parser so I can build a solid HTTP client library!

uptycs-nishant commented:
@theopolis , I am done with code changes for this PR. I believe, we can address code changes pertaining to the user code ( code using http client) can be addressed in a separate PR.
